cpp/inc/bond/core/traits.h

```diff
@@ -168,6 +175,40 @@ template <typename T> struct
 is_type_alias
     : is_object<typename aliased_type<T>::type> {};
 
+
+// is_reader
+template <typename Input, typename T = void, typename Enable = void> struct
+is_reader
+    : false_type {};
+
+template <typename Input, typename T> struct
+is_reader<Input&, T>
```

Comment:
You cannot use any such traits in partial specializations, because compiler won't be able to deduce the type (e.g. `is_reader<typename remove_reference<T>::type, T>`). So either (1) you specialize for individual `T` which is not a reference or `const/volatile` and consumers are explicitly using `remove_reference/remove_cv`, or (2) we add such partial specializations for `T&`, `const/volatile T` which falls back to `T` like that is done above so that users do not have to worry about that (assuming this trat should not behave differently for `T` vs `T&` etc.).